File: src/drivers/device/keypad.cpp
/**
 * This file is a part of ThetaGP.
 *
 * ThetaGP is free software: you can redistribute it
 * and/or modify it under the terms of the GNU General
 * Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * ThetaGP is distributed in the hope that it will be
 * useful, but WITHOUT ANY WARRANTY; without even the
 * implied warranty of MERCHANTABILITY or FITNESS FOR A
 * PARTICULAR PURPOSE. See the GNU General Public License
 * for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program.
 *
 * If not, see <https://www.gnu.org/licenses/>.
 */

#include "BoardConfig.h"
#include "conf/ThetaGP_Config.h" // THETAGP_CFG_KEY_TOGGLE_EN, the keypad toggle switch

#include "drivers/device/keypad.h"

#include "drivers/device/systimer.h"
#include "drivers/peripherals/gpio.h"
#include "drivers/peripherals/nvic.h"
#include "drivers/peripherals/peripheralsmgr.h"
#include "drivers/peripherals/systick.h"

#include "utils/atomic.h"
#include "utils/log/log.h"

namespace ThetaGP::Drivers::Device {

using namespace Peripheral::GPIO;
using namespace Peripheral::TIMER;

Keypad::Keypad() : Device("keypad") {}

void Keypad::init() {
  if (_initialized)
    return;

  initPins();

  _readInput = &Keypad::readInputScanMatrix;

  auto setupTimer = [this](HardwareTimer &timer,
                           Peripheral::TIMER::Instance instance) {
    timer.config(instance, KeypadConfig::DEFAULT_SCAN_FREQ);
    timer.setCallback(
        [](void *self) {
          auto *keypad = static_cast<Keypad *>(self);
          if (keypad)
            keypad->scanCallback();
        },
        this);
    timer.init();
    return timer.isInitialized();
  };

  setupTimer(_scanTimer, Peripheral::PeripheralsManager::getInstance().reservedTimer());

  _scanTimer.start();
  _initialized = true;
}

void Keypad::scanCallback() {
  // Minimum measurement point: one read of the device layer's cycle counter at
  // each end of the callback. SystemTimer is how the rest of the device layer
  // takes time (the scheduler reads its clock the same way), and the keypad is a
  // device, so it takes the time from the device-layer timer rather than from a
  // peripheral register. The counter form is the one to use here: its tick is
  // one CPU cycle, so a callback that lasts ten microseconds is measured to the
  // cycle, while the microsecond clock would quantize it to ten steps and cost
  // an atomic block and a division per read. Both ends are one read plus one
  // call, and the host converts to microseconds where a conversion is wanted.
  const uint32_t startCycles = SystemTimer::getInstance().getCycleCounter();

  uint32_t mask = 0;
  (this->*_readInput)(&mask);

  // Per-scan decision, one key at a time, from the raw mask this scan read.
  // A sample that agrees with a key's committed state clears both of its runs,
  // so a run of the opposite level counts only while it stays unbroken and a
  // burst shorter than the threshold that guards that direction leaves no trace
  // at all. A run that reaches its threshold commits the level it counted and
  // then starts over.
  uint32_t committedMask = 0;

  for (size_t i = 0; i < MAX_KEYS; i++) {
    KeySampler &s = _samplers[i];
    const bool pressed = (mask & (1U << i)) != 0;
    const bool stable = (s.stableState == KeyState::Pressed);

    if (pressed == stable) {
      s.pressRun = 0;
      s.releaseRun = 0;
    } else if (pressed) {
      s.releaseRun = 0;
      if (++s.pressRun >= KeypadConfig::PRESS_SAMPLES) {
        s.stableState = KeyState::Pressed;
        s.pressRun = 0;
      }
    } else {
      s.pressRun = 0;
      if (++s.releaseRun >= KeypadConfig::RELEASE_SAMPLES) {
        s.stableState = KeyState::Released;
        s.releaseRun = 0;
      }
    }

    if (s.stableState == KeyState::Pressed) {
      committedMask |= (1U << i);
    }
  }

  // The committed mask is assembled whole and published once, so a reader sees
  // either the previous mask or this one, never a mixture of two keys.
  _pressedMask = committedMask;

  // Raw cycles, kept unconverted so the counters hold the measurement at full
  // resolution; the readout converts. The pair of stamps costs a few cycles and
  // is not subtracted out — at this scale the measurement is about five thousand
  // cycles, so the cost is well under a tenth of a percent of it.
  const uint32_t elapsedCycles = SystemTimer::getInstance().getCycleCounter() - startCycles;
  _scanCyclesLast = elapsedCycles;
  _scanCyclesSum += elapsedCycles;
  if (elapsedCycles > _scanCyclesMax) {
    _scanCyclesMax = elapsedCycles;
  }
  _scanCyclesCount++;
}

void Keypad::getScanStats(ScanStats &out) const {
  ATOMIC_BLOCK(NVIC_PRIO_MAX) {
    out.count = _scanCyclesCount;
    out.last_cycles = _scanCyclesLast;
    out.max_cycles = _scanCyclesMax;
    out.sum_cycles = _scanCyclesSum;
  }
}

void Keypad::readInputScanMatrix(uint32_t *mask) {
  const bool activeLow = (_active == KeypadConfig::Active::Low);
  const PinState driveState = activeLow ? PinState::Reset : PinState::Set;
  const PinState idleState = activeLow ? PinState::Set : PinState::Reset;
  const PinState senseState = driveState;

  for (size_t d = 0; d < DRIVE_PIN_NUM; d++) {
    Gpio driveGpio(_drivePins[d]);
    driveGpio.write(driveState);

    for (uint32_t i = 0; i < KeypadConfig::GPIO_STABILIZE_DELAY_CYCLES; i++) {
      __NOP();
    }

    for (size_t s = 0; s < SENSE_PIN_NUM; s++) {
      Gpio senseGpio(_sensePins[s]);
      if (senseGpio.read() == senseState) {
        const uint8_t keyId = getKeyId(d, s);
        if (isValidKey(keyId)) {
          *mask |= (1U << keyId);
        }
      }
    }

    driveGpio.write(idleState);
  }
}

void Keypad::initPins() {
#if defined(BDCFG_KEYPAD_DRIVE_IO_LIST) && defined(BDCFG_KEYPAD_SENSE_IO_LIST)
  const PinDesc drivePinsTmp[] = {BDCFG_KEYPAD_DRIVE_IO_LIST};
  const PinDesc sensePinsTmp[] = {BDCFG_KEYPAD_SENSE_IO_LIST};

  const bool activeLow = (_active == KeypadConfig::Active::Low);
  const PinState idleState = activeLow ? PinState::Set : PinState::Reset;
  const Pull sensePull = activeLow ? Pull::PullUp : Pull::PullDown;

  for (size_t i = 0; i < DRIVE_PIN_NUM; i++) {
    _drivePins[i] = drivePinsTmp[i];
    Gpio gpio(_drivePins[i]);
    gpio.config(Mode::OutputPushPull, Pull::NoPull, Speed::VeryHigh);
    gpio.init();
    gpio.write(idleState);
  }

  for (size_t i = 0; i < SENSE_PIN_NUM; i++) {
    _sensePins[i] = sensePinsTmp[i];
    Gpio gpio(_sensePins[i]);
    gpio.config(Mode::Input, sensePull, Speed::High);
    gpio.init();
  }
#endif
}

uint32_t Keypad::getPressed() const {
  UNUSED(this);
#if THETAGP_CFG_KEY_TOGGLE_EN
  // Test hook: flip one key bit per read. The gamepad tick reads this once, so
  // the mask differs from the previous report on every tick. A free-running
  // counter is used rather than a timer bit because the tick period is an even
  // number of microseconds, which would keep a micros()-derived bit's parity
  // from tick to tick.
  static uint32_t s_readCount = 0;
  return _pressedMask ^ ((++s_readCount & 1U) ? 0x1U : 0x2U);
#else
  return _pressedMask;
#endif
}

bool Keypad::isKeyPressed(uint8_t keyId) const {
  if (keyId >= MAX_KEYS)
    return false;
  return (_pressedMask & (1U << keyId)) != 0;
}

} // namespace ThetaGP::Drivers::Device

File: scripts/config_lib/generators/keypad.lua
--[[
# This file is a part of ThetaGP.
#
# ThetaGP is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# ThetaGP is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/>.
]]

-- scripts/config_lib/generators/keypad.lua

local utils = require("utils")

local M = {}

local DRIVE_MODE_MAP = {
    scan_matrix = "ScanMatrix",
    io_direct = "IODirect",
    spi_74hc165 = "SpiDriven74HC165",
}

local ACTIVE_MODE_MAP = {
    none = "None",
    low = "Low",
    high = "High",
}

function M.generate(keypad_config)
    local lines = {}

    if not keypad_config then
        return lines
    end

    if keypad_config.drive_mode then
        local mode_value = DRIVE_MODE_MAP[keypad_config.drive_mode:lower()] or ""
        table.insert(lines, string.format(
            '#define %-28s %s',
            'KEYPAD_DRIVE_MODE',
            "KeypadConfig::Mode::" .. mode_value
        ))
    end

    if keypad_config.active_mode then
        local active_value = ACTIVE_MODE_MAP[keypad_config.active_mode:lower()] or "None"
        table.insert(lines, string.format(
            '#define %-28s %s',
            'KEYPAD_ACTIVE_MODE',
            "KeypadConfig::Active::" .. active_value
        ))
    else
        table.insert(lines, string.format(
            '#define %-28s %s',
            'KEYPAD_ACTIVE_MODE',
            "KeypadConfig::Active::None"
        ))
    end

    local drive_mode = keypad_config.drive_mode:lower()

    if drive_mode == "scan_matrix" then
        if keypad_config.drive_pins and #keypad_config.drive_pins > 0 then
            table.insert(lines, string.format(
                '#define %-28s %d',
                'KEYPAD_DRIVE_PIN_NUM',
                #keypad_config.drive_pins
            ))
            table.insert(lines, utils.generate_pin_array_macro(
                'KEYPAD_DRIVE_IO_LIST',
                keypad_config.drive_pins
            ))
        end

        if keypad_config.sense_pins and #keypad_config.sense_pins > 0 then
            table.insert(lines, string.format(
                '#define %-28s %d',
                'KEYPAD_SENSE_PIN_NUM',
                #keypad_config.sense_pins
            ))
            table.insert(lines, utils.generate_pin_array_macro(
                'KEYPAD_SENSE_IO_LIST',
                keypad_config.sense_pins
            ))
        end

        if keypad_config.key_map then
            local drive_num = #keypad_config.drive_pins
            local sense_num = #keypad_config.sense_pins
            local total_keys = drive_num * sense_num

            table.insert(lines, "")
            table.insert(lines, "#define KEYPAD_KEY_MAP \\")

            local max_index = 0
            local item_count = 0
            local total_items = drive_num * sense_num

            for d = 1, drive_num do
                for s = 1, sense_num do
                    item_count = item_count + 1
                    local value = keypad_config.key_map[d][s]
                    if value == nil then
                        value = 0xFF  -- KEYPAD_NO_KEY
                    end
                    if value ~= 0xFF and value > max_index then
                        max_index = value
                    end

                    if item_count < total_items then
                        table.insert(lines, string.format("    %3d, \\", value))
                    else
                        table.insert(lines, string.format("    %3d", value))
                    end
                end
            end

            table.insert(lines, "")
            table.insert(lines, "")
            table.insert(lines, string.format("#define %-28s %d", 'KEYPAD_MAX_KEY_INDEX', max_index))
            table.insert(lines, string.format("#define %-28s %d", 'KEYPAD_MASK_ARRAY_SIZE', math.floor((max_index + 32) / 32)))
            table.insert(lines, string.format("#define %-28s %d", 'KEYPAD_KEY_MAP_SIZE', total_keys))
        end

    elseif drive_mode == "io_direct" then
        if keypad_config.direct_pins and #keypad_config.direct_pins > 0 then
            table.insert(lines, string.format(
                '#define %-28s %d',
                'KEYPAD_DIRECT_PINS_NUM',
                #keypad_config.direct_pins
            ))
            table.insert(lines, utils.generate_pin_array_macro(
                'KEYPAD_DIRECT_PINS',
                keypad_config.direct_pins
            ))
        end

    elseif drive_mode == "spi_74hc165" then
        if keypad_config.spi_chips then
            table.insert(lines, string.format(
                '#define %-28s %d',
                'KEYPAD_SPI_CHIPS',
                keypad_config.spi_chips
            ))
        end
    end

    if keypad_config.button_map and #keypad_config.button_map > 0 then
        table.insert(lines, "")
        table.insert(lines, "#define KEYPAD_BUTTON_MAP \\")

        local count = #keypad_config.button_map
        for i, v in ipairs(keypad_config.button_map) do
            local mask_name = "GAMEPAD_MASK_" .. v:upper()
            if i < count then
                table.insert(lines, string.format("    %-20s, \\", mask_name))
            else
                table.insert(lines, string.format("    %-20s", mask_name))
            end
        end

        table.insert(lines, "")
    end

    return lines
end

return M
